### DISCUSSION TITLE
custom dns api in docker

### DISCUSSION THREAD

**Discussion by svde** [acmesh-official/acme.sh#4962] [Q&A] [answered]
Quick question: where am I supposed to place the custom dns api script in case of docker, and how am I supposed to call it?

It's complaining: "Can not find dns api hook for: : dns_solidserver", 

Calling acme.sh like this:
`podman run --rm  -it -v "$(pwd)/out":/acme.sh --net=host neilpang/acme.sh --issue --dns dns_solidserver -ak 4096 -k 4096 -d test.*********.com`

Script is in ~/out/dns_solidserver.sh, exactly the same script we're using on other (non-docker) acme.sh systems, where it's in /usr/share/acme.sh/dnsapi/dns_solidserver.sh.

**Comment by neilpang** [comment]
you can use `--dns /path/to/your/script.sh`  to specify the full path of the custom api.

**Reply by svde** [reply]
Doesn't work for me:

```
❯ podman run --rm  -it -v "$(pwd)/out":/acme.sh --net=host neilpang/acme.sh --issue --dns /acme.sh/dns_solidserver.sh -ak 4096 -k 4096 -d test.****.com
[Fri Jan 26 10:01:34 UTC 2024] Using CA: https://acme-v02.api.letsencrypt.org/directory
[Fri Jan 26 10:01:34 UTC 2024] Single domain='test.****.com'
[Fri Jan 26 10:01:34 UTC 2024] Getting domain auth token for each domain
[Fri Jan 26 10:01:36 UTC 2024] Getting webroot for domain='test.****.com'
[Fri Jan 26 10:01:36 UTC 2024] Verifying: test.****.com
mkdir: cannot create directory ‘/acme.sh/dns_solidserver.sh’: Not a directory
/root/.acme.sh/acme.sh: line 4945: can't create /acme.sh/dns_solidserver.sh/.well-known/acme-challenge/CCBp1DXzELaNhpwPsoFyBgNJc6uSpeOPsJNlbpmGfRA: nonexistent directory
```

the file is there:

```
❯ podman run --rm  -it -v "$(pwd)/out":/acme.sh --net=host neilpang/acme.sh /bin/ls /acme.sh/
account.conf  ca  dns_solidserver.sh  http.header  test.****.com
```

**Comment by neilpang** [comment, marked as answer]
Correction: `--dns` takes a hook NAME (it must start with `dns_`), not a file path -- a path is treated as a webroot directory, which is where the mkdir error came from. And the hook is searched in the acme.sh script dir, not in the mounted config home (`/acme.sh`). Mount the custom script into the image's dnsapi folder and call it by name:

```
podman run --rm -v "$(pwd)/out":/acme.sh \
  -v "$(pwd)/out/dns_solidserver.sh":/acmebin/dnsapi/dns_solidserver.sh \
  --net=host docker.io/neilpang/acme.sh \
  --issue --dns dns_solidserver -k 4096 -d test.example.com
```

Current images use `/acmebin` as the script dir; older images used `/root/.acme.sh`.
